// socketsyncr/client.js
MODULE_NAME = 'SocketSyncr (Client)';
var log = function(msg) {
	console.log(MODULE_NAME + ': '+msg)
}
log('<-- Loading...')

shared = require('./shared')


var SyncClient = exports.SyncClient = function(host,options) {
	
	
	var self = this;
	
	this.syncables = {};
	
	this.socket = new io.Socket(host,options);
	
	this.connect = function() {
		self.socket.connect();
	}
	
	/* Message handler */
	this.socket.on('message', function(message) {
		
		log('Received from server:')
		//log(message);
		log('[truncated]')
		
		obj = JSON.parse(message);
		
		log(JSON.stringify(obj[shared.SYNCABLE_EVENT_]))
		
		/* Handler generic events */
		if (obj[shared.SYNC_EVENT_]) {
			
			log("Got 'normal' sync event.")
			
			event = obj[shared.SYNC_EVENT_].shift() // First item is event
			listener = self.once[event]
			if (listener) {
				delete self.once[event]
			} else {
				listener = self.listeners[event]
			}
			if (listener) {
				listener.apply(listener,obj[shared.SYNC_EVENT_]) // Rest of items are args for event
			}
			
		/* Handle syncable objects events */
		} else if (obj[shared.SYNCABLE_EVENT_]) {
			
			log("Got 'syncable' obj event")
			
			sync_event = obj[shared.SYNCABLE_EVENT_];
			listener = self.server_syncable_listeners[sync_event];
			listener(obj['syncable'],obj['argument'])
			
		}
	})
	
	
	
	/* GENERIC SYNC EVENTS */
	
	this.listeners = {}
	this.once = {}
	this.on = function(event,listener) {
		self.listeners[event] = listener;
	}
	
	this.once = function(event,listener) {
		self.once[event] = listener;
	}
	
	this.emit = function(event,args) {
		args.unshift(event)
		obj = {}; obj[shared.SYNC_EVENT_] = args;
		self.socket.send(JSON.stringify(obj))
	}
	
	
	
	
	/* SYNCABLES */
	
	this.syncables = {}
	this.syncable_constructors = {}
	
	
	this.register = function(syncable) {
		
		if (!syncable) log("Warning: trying to register '"+syncable+"'.")
		
		self.syncable_constructors[syncable.constructor.toString()] = syncable.constructor;
		
		from_server = Boolean(syncable[shared.SYNC_ID_] !== null || syncable[shared.SYNC_ID_] !== undefined)
		if (!from_server) {
		
			// We'll use a temporary ID for this syncable until we get one from the server
			temp_id = shared.TEMP_ID_PREFIX_ + Math.random()
			self.syncables[temp_id] = syncable;
			self.syncable[shared.SYNC_ID_] = temp_id
		
			self.on_server_syncable_event(shared.SERVER_SET_ID_,function(dummy_syncable,arguments) { // <-- Shit, we need to ensure we get the right server id for the right syncable
				// The server gave us an authoritative ID for this object. Save it.
			
				syncable = syncables[arguments[0]]; // <-- arg0 is temp_id
				delete syncables[arguments[0]];
				syncable[shared.SYNC_ID_] = arguments[1]; // <-- arg1 is authoritative id
				syncables[arguments[1]] = syncable;
			})
		
			/* Ask the server for an authoritative id for this syncable */
			self.socket.send(
				shared.full_syncable_event_message(shared.NEW_SYNCABLE_,syncable)
			)
		} else { // is from the server
			self.syncables[shared.SYNC_ID_] = syncable;
		}
		
		syncable.on('change',function() {
			self.socket.send(
				shared.lite_syncable_event_message(shared.SYNCABLE_CHANGED_,syncable)
			)
		})
	
	}
	
	this.deregister = function(syncable) {
		throw 'Not yet implemented!';
	}
	
	
	this.server_syncable_listeners = {}
	this.on_server_syncable_event = function(event,listener) {
		self.server_syncable_listeners[event] = listener;
	}
	
	this.on_server_syncable_event(shared.NEW_SYNCABLE_,function(syncable_json,args) {
		
		log('New syncable event! ')
		
		syncable = shared.objectify_syncable_json(syncable_json)
		
		log('Built syncable ' + JSON.stringify(syncable))
		
		self.register(syncable);
		
	})
	
	this.on_server_syncable_event(shared.SYNCABLE_CHANGED_, function(syncable_json, args) {
		
		log('Syncable changed event!')
		
		syncable = self.syncables[syncable_json[shared.SYNC_ID_]];
		if (!syncable)
			log('Got unknown syncable!')
		syncable.changed(syncable_json);
		
	});
	
}
